azure-pipelines.yml

```diff
@@ -46,3 +144,149 @@ jobs:
+    displayName: 'Configure git'
+    env:
+      GH_NAME: $(gh.name)
+      GH_EMAIL: $(gh.email)
+
+  - script: |
+      pip install -e .
+    displayName: Install pyvista
+
+  - script: |
+      .ci/setup_headless_display.sh
+      # python .ci/pyvista_test.py  # for debug
+    displayName: Install headless display
+
+  - script: |
+      sudo apt-get install python3-tk
+      pip install -r requirements_docs.txt
+    displayName: 'Install dependencies'
+
+  - script: |
+      EXAMPLES_PATH=$(python -c "import pyvista; print(pyvista.EXAMPLES_PATH)")
+      mkdir -p EXAMPLES_PATH
+      echo "##vso[task.setvariable variable=EXAMPLES_PATH]$EXAMPLES_PATH"
+      PYVISTA_VERSION=$(python -c "import io; exec(io.open('pyvista/_version.py').read()); print(__version__)")
+      echo "##vso[task.setvariable variable=PYVISTA_VERSION]$PYVISTA_VERSION"
+      echo $EXAMPLES_PATH
+    displayName: 'Store location of pyvista downloads cache and version'
+  
+  - task: Cache@2
+    inputs:
+      key: 'PYVISTA_EXAMPLES_CACHE | "$(PYVISTA_VERSION)"'
+      path: $(EXAMPLES_PATH)
+    displayName: Build pyvista examples cache
+
+  - script: |
+      make -C docs html
+    displayName: 'Build documentation'
+
+  - script: |
+      make -C docs doctest
+    displayName: 'Test documentation code snippets'
+
+  - script: |
+      make -C docs html
+    displayName: 'Update figures from doctest'
+
+  # Push example notebooks to pyvista-examples
+  - script: |
+      pip install cookiecutter
+      cookiecutter -f --no-input --config-file ./docs/pyvista-binder-config.yml -o .. https://github.com/pyvista/cookiecutter-pyvista-binder.git;
+      rm -rf ../pyvista-examples/notebooks/
+      cd docs
+      find examples -type f -name '*.ipynb' | cpio -p -d -v ../../pyvista-examples/
+      cd ../../pyvista-examples/
+      git init
+      git add .
+      git commit -am "Azure CI commit ref $(Build.SourceVersion)"
+      REMOTE="https://${GH_TOKEN}@github.com/pyvista/pyvista-examples"
+      git remote add origin ${REMOTE}
+      git push -uf origin master
+      cd $BUILD_SOURCESDIRECTORY
+    displayName: 'Upload binder examples'
+    condition: and(succeeded(), contains(variables['Build.SourceBranch'], 'refs/tags/'))
+    env:
+      GH_TOKEN: $(gh.token)
+
+  # upload documentation to pyvista-docs only on master
+  - script: |
+      git clone --depth 1 https://${GH_TOKEN}@github.com/pyvista/pyvista-docs.git
+      cd pyvista-docs
+      git gc --prune=now
+      git remote prune origin
+      rm -rf *
+      cp -r $BUILD_SOURCESDIRECTORY/docs/_build/html/* .
+      cp $BUILD_SOURCESDIRECTORY/docs/README.md .
+      touch .nojekyll
+      echo "docs.pyvista.org" >> CNAME
+      git add .
+      git commit -am "Azure CI commit ref $(Build.SourceVersion)"
+      git push 
+    displayName: Upload Documentation to pyvista-docs
+    env:
+      GH_TOKEN: $(gh.token)
+    condition: and(succeeded(), contains(variables['Build.SourceBranch'], 'refs/tags/'))
```

Comment:
Will this only deploy on tags? Can we deploy on master a few times before a tag to make sure everything is working?

Reply:
Why not deploy to a different branch (e.g. `develop`) merged with master and then deploy to `gh-pages` when tagged?

Reply:
Pushing to develop for both the binder examples and html docs when merging to master.  That work for you?

Reply:
sounds good to me